File: Easy/700_Search_In_a_Binary_Search_Tree/Solution.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def searchBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        curr = root

        while (curr != None):
            if (curr.val == val):
                return curr
            elif (curr.val < val):
                curr = curr.right
            else:
                curr = curr.left
        
        return None


if __name__ == "__main__":
    root = TreeNode(4)

    root.left = TreeNode(2)
    root.right = TreeNode(7)

    root.left.left = TreeNode(1)
    root.left.right = TreeNode(3)

    solution = Solution()

    # test cases 1
    print(solution.searchBST(root, 2).val)

    # test cases 2
    print(solution.searchBST(root, 5))

File: Easy/700_Search_In_a_Binary_Search_Tree/Solution.java


public class Solution {
    public TreeNode searchBST(TreeNode root, int val) {
        TreeNode curr = root;

        while (curr != null) {
            if (curr.val == val) {
                return curr;
            } else if (curr.val < val) {
                curr = curr.right;
            } else {
                curr = curr.left;
            }
        }

        return null;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(4);

        root.left = new TreeNode(2);
        root.right = new TreeNode(7);

        root.left.left = new TreeNode(1);
        root.left.right = new TreeNode(3);

        Solution solution = new Solution();

        // test cases 1
        TreeNode result1 = solution.searchBST(root, 2);
        System.out.println(result1.val);

        // test cases 2
        TreeNode result2 = solution.searchBST(root, 5);
        System.out.println(result2);

    }
}



// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}